src/obj/keymgmt.c

```diff
@@ -0,0 +1,928 @@
+    int rc = RET_OSSL_ERR;
+
+    /* is BN ?*/
+    if (attr != CKA_MODULUS && attr != CKA_PUBLIC_EXPONENT) {
+        return rc;
+    }
+
+    x1 = p11prov_obj_get_attr(key1, attr);
+    x2 = p11prov_obj_get_attr(key2, attr);
+
+    if (!x1 || !x2) {
+        return rc;
+    }
+
+    bx1 = BN_native2bn(x1->pValue, x1->ulValueLen, NULL);
+    bx2 = BN_native2bn(x2->pValue, x2->ulValueLen, NULL);
+
+    if (BN_cmp(bx1, bx2) == 0) {
+        rc = RET_OSSL_OK;
+    }
+
+    BN_free(bx1);
+    BN_free(bx2);
+
+    return rc;
+}
+
+static int cmp_attr(P11PROV_OBJ *key1, P11PROV_OBJ *key2,
+                    CK_ATTRIBUTE_TYPE attr)
+{
+    CK_ATTRIBUTE *x1, *x2;
+
+    x1 = p11prov_obj_get_attr(key1, attr);
+    x2 = p11prov_obj_get_attr(key2, attr);
+    if (!x1 || !x2) {
+        return RET_OSSL_ERR;
+    }
+    if (x1->ulValueLen != x2->ulValueLen) {
+        return RET_OSSL_ERR;
+    }
+    if (memcmp(x1->pValue, x2->pValue, x1->ulValueLen) != 0) {
+        return RET_OSSL_ERR;
+    }
+    return RET_OSSL_OK;
+}
+
+static int cmp_public_key_values(P11PROV_OBJ *pub_key1, P11PROV_OBJ *pub_key2)
+{
+    int ret;
+
+    switch (pub_key1->data.key.type) {
+    case CKK_RSA:
+        /* pub_key1 pub_key2 could be CKO_PRIVATE_KEY here but
+         *  nevertheless contain these two attributes
+         */
+        ret = cmp_bn_attr(pub_key1, pub_key2, CKA_MODULUS);
+        if (ret == RET_OSSL_ERR) {
+            break;
+        }
+        ret = cmp_bn_attr(pub_key1, pub_key2, CKA_PUBLIC_EXPONENT);
+        break;
+    case CKK_EC:
+    case CKK_EC_EDWARDS:
+        ret = cmp_attr(pub_key1, pub_key2, CKA_P11PROV_PUB_KEY);
+        break;
+    case CKK_ML_DSA:
+        ret = cmp_attr(pub_key1, pub_key2, CKA_VALUE);
+        break;
+    case CKK_ML_KEM:
+        ret = cmp_attr(pub_key1, pub_key2, CKA_VALUE);
+        break;
+    default:
+        ret = RET_OSSL_ERR;
+    }
+
+    return ret;
+}
+
+static int match_key_with_cert(P11PROV_OBJ *priv_key, P11PROV_OBJ *pub_key)
+{
+    P11PROV_OBJ *cert;
+    CK_ATTRIBUTE attrs[2] = { 0 };
+    CK_ATTRIBUTE *x;
+    int num = 0;
+    int ret = RET_OSSL_ERR;
+
+    cert = p11prov_obj_find_associated(priv_key, CKO_CERTIFICATE);
+    if (!cert) {
+        P11PROV_raise(priv_key->ctx, CKR_GENERAL_ERROR,
+                      "Could not find associated certificate object");
+        return RET_OSSL_ERR;
+    }
+
+    switch (pub_key->data.key.type) {
+    case CKK_RSA:
+        attrs[0].type = CKA_MODULUS;
+        attrs[1].type = CKA_PUBLIC_EXPONENT;
+        num = 2;
+        break;
+    case CKK_EC:
+    case CKK_EC_EDWARDS:
+        attrs[0].type = CKA_P11PROV_PUB_KEY;
+        num = 1;
+        break;
+    }
+
+    ret = get_attrs_from_cert(cert, attrs, num);
+    if (ret != CKR_OK) {
+        P11PROV_raise(priv_key->ctx, ret,
+                      "Failed to get public attrs from cert");
+        ret = RET_OSSL_ERR;
+        goto done;
+    }
+
+    switch (pub_key->data.key.type) {
+    case CKK_RSA:
```

Comment:
TODO ML-DSA certificate comparsion missing

Reply:
same as bug above we need a ML-DSA cert and some tests to make sure this works.